ADDED: ability to filter coverage based on unified diff file

rburtnykx opened the pull request:
Hi,

In this PR want to propose ability to filter coverage according to unified diff file. 

So simple use case is generating diff for pull request, running fastcov with diff + lcov_to_cobertura for CI coverage based on changes. Or diff for pull request, fastcov with diff + genhtml to get changes coverage report.
Currently see no good way to generate changes html coverage report, because usually that is one html file, which become very big and hard to open using browser.
Also good thing is that fastcov can be used as a intermediate step between lcov+genhtml  or lcov + lcov_to_cobertura if someone need to get changes coverage report even if GCC version < 9. 

Please, let me know if this change is reasonable, so I will add some unittests to check diff parsing and functional to check overall scenario.

RPGillespie6 commented:
Can you really perform manipulations on a report from a diff file without losing correctness? I don't think this method will yield correct results.

For example, take the following sample diff file:
```
--- lao	2002-02-21 23:30:39.942229878 -0800
+++ tzu	2002-02-21 23:30:50.442260588 -0800
@@ -1,4 +1,3 @@
int x = 0;
-x = 17;
if (x == 17)
    printf("hello world");
```

Would this correctly flag lines 2 and 3 as having changed coverage, because the if condition is no longer being met?

To guarantee correctness you need to generate 2 fastcov reports - one generated with diffs applied, one without. Once you have the two reports, you can then subtract one from the other to determine change in coverage. 

I'm still not wrapping my head around this feature, could you give a simple example of what you are trying to accomplish?

rburtnykx commented:
So main idea is pretty the same as codecov/patch did in this PR. Sometimes it is useful to know only coverage for changed code. That includes changed and added lines. F.e. in this PR I want to know coverage for my changes, that is fastcov.py file and next lines: 63-186, 837-838, 914-917. So I added 130 lines and only 11 of them is covered, so coverage for my changes is near 8%. 
Diff file is like:
@@ -60,6 +60,130 @@ class FastcovFormatter(logging.Formatter):
         log_message = super(FastcovFormatter, self).format(record)
         return "[{:.3f}s] {}".format(stopwatch(), log_message)

+class DiffParser(object):
+    def _refinePaths(self, diff_metadata, diff_base_dir):
+        diff_metadata.pop('/dev/null', None)
+ ...
@@ -710,6 +834,8 @@ def parseArgs():
 ...
     parser.add_argument('-f', '--gcda-files',   dest='coverage_files',  nargs="+", metavar='', default=[], help='Filter: Specify exactly which gcda or gcno files should be processed. Note that specifying gcno causes both gcno and gcda to be processed.')
     parser.add_argument('-E', '--exclude-gcda', dest='excludepre',  nargs="+", metavar='', default=[], help='Filter: Exclude gcda or gcno files from being processed via simple find matching (not regex)')
+    parser.add_argument('-u', '--diff-filter', dest='diff_file', default='', help='Unified diff file with changes which will be included into final report')
+    parser.add_argument('-ub', '--diff-base-dir', dest='diff_base_dir', default='.', help='Base directory for sources in unified diff file, usually repository dir')
...

In other words, some project f.e. has next rule: all changed code should be covered at least on 80%. Then we need to determine which lines was changed from whole project (modified or added), and which of them are covered. Information about which lines from some source file was changed is located in diff file. Lines are marked by '+'. So we generate coverage for whole project, then get all line numbers from diff file, and use them as "include" filter -> only they will be present in final report. SO instead of total coverage we will receive only coverage for changes. 

In your example you have only 1 deleted line, so it means 0 changed lines need to be covered and 0 of them are covered.
But lets say, you  changed 17 to 10:
--- lao	2002-02-21 23:30:39.942229878 -0800
+++ tzu	2002-02-21 23:30:50.442260588 -0800
@@ -1,4 +1,4 @@
 int x = 0;
-x = 17;
+x = 10;
 if (x == 17)
     printf("hello world");

Your file looks like next:
 int x = 0;
 x = 10;
 if (x == 17)
     printf("hello world");

So only 2nd line will be highlighted, and coverage will  be 0/1 or 1/1, depends on was that line covered previously or not.
So main idea of that diff file is to use it as a filter which # of lines should be included.

RPGillespie6 commented:
My point though is that the results are not necessarily correct and can/will be misleading. To use your example:

Say you have:
```
--- lao	2002-02-21 23:30:39.942229878 -0800
+++ tzu	2002-02-21 23:30:50.442260588 -0800
@@ -1,4 +1,4 @@
 int x = 0;
-x = 17;
+x = 10;
 if (x == 17) {
     printf("hello world");
     printf("hello world2");
     printf("hello world3");
}
```

Now file looks like this:
```
int main()
{
    int x = 0;
    x = 10;
    if (x == 17) {
        printf("hello world");
        printf("hello world2");
        printf("hello world3");
    }
    return 0;
}
```

And say your unit test is:
```
main()
```

Before diff is applied, there is 100% coverage. After diff is applied, coverage drops by 3 lines:
```
        printf("hello world");
        printf("hello world2");
        printf("hello world3");
```

The added line, `x = 10;` is actually still covered by unit tests. so it's misleading to say 0/1 new lines are not covered. The correct diff would be that coverage dropped by 3 lines.

rburtnykx commented:
Yes, about applying diff file to coverage are are absolutely right. This cannot be done easily, and the best way is to apply it to sources and rerun unittests. But change is not about applying diff to coverage. It is more about filtering coverage including only changed lines there. 

So report generated by genhtm will be like
```
int main()
{
    int x = 0;
    **x = 10;**
    if (x == 17) {
        printf("hello world");
        printf("hello world2");
        printf("hello world3");
    }
    return 0;
}
```
```
**int main()
{
    int x = 0;
    x = 10;
    if (x == 17) {
        printf("hello world");
        printf("hello world2");
        printf("hello world3");
    }
    return 0;
}**
```
Result will looks like report, generated by codecov/patch check to this PR
https://codecov.io/gh/RPGillespie6/fastcov/compare/2cad8ec4415db088eaf9a7172428f98bf5c20ee7...5bf89ee780d1045bfd95b879fbaaa73390c2182b/diff

RPGillespie6 commented:
I see. So you are already passing in the diff coverage report and just making the lines different colors based on the unified diff?

A few questions:
- Would it be worth using something like [unidiff](https://github.com/matiasb/python-unidiff) to avoid having to test the homegrown parser? I still want people to be able to run fastcov.py directly without needing a bunch of 3rd party stuff, but I would be okay with:

```
 if args.diff_file:
        import unidiff
        logging.info("Filtering according to {} file".format(args.diff_file))
        filterByDiff(args.diff_file, args.diff_base_dir, fastcov_json, args.fallback_encodings)
```

- Would it be worth breaking this out into a separate CLI tool in the same repository? It doesn't really use any of fastcov's core functionality as far as I can tell.

There's pros and cons I suppose. On the one hand it's nice to have one tool do it all (swiss army knife). But also there is more testing overhead, and I'm happy to add separate CLI utilities to the repository that are not as rigorously tested. Thoughts?

RPGillespie6 commented:
I'm fine with this going into fastcov main file as long as you unit tests to cover the class (and one functional test to test CLI argument) like you mentioned. Feel free to create a directory containing unified diff files for testing.

rburtnykx commented:
> I see. So you are already passing in the diff coverage report and just modifying it so it only includes the lines touched in the unified diff file?
Yes, exactly. That is smth like scanExclusionMarkers logic, but that is about exclude lines from coverage and diff file is about include only. So that is used for PR validation and make sure changes are covered by some %, in this case all except changed lines are expected to be removed from report.

> Would it be worth using something like unidiff to avoid having to test the homegrown parser? I still want people to be able to run fastcov.py directly without needing a bunch of 3rd party stuff, but I would be okay with:
I think it can be done. So yes, very good thing in fastcov is that it has no dependencies, but unidiff is not very big and consist from 3 py files. On other hand parser is not very complicated and not sure that will be very hard to cover by test. So let me know which one is the best, so i will add tests or usage of unidiff. 

>Would it be worth breaking this out into a separate CLI tool in the same repository. 
Generally I'm OK with that, but it will use all combine logic from fastcov. At least lcov/json parser and lcov/json writer needed here. So it can be separate CLI tool or just additional filter, like exclude, include, scan_exclusion_markers. 

>I'm fine with this going into fastcov main file as long as you write unit tests to cover the class (and one functional test to test CLI argument) like you mentioned. Feel free to create a directory containing unified diff files for testing.
>
>I might ping you with an @ mention though if issues are opened against this feature in the future ;)

Sure, I will add needed tests to cover this feature, thanks. And yes, feel free to ping me if any issue will be there: ) (I hope there will be  opened issues against this feature, that will mean not only me found this feature usefull :) )

RPGillespie6 commented:
> So let me know which one is the best, so i will add tests or usage of unidiff.

If you decide to make a separate CLI tool in the same repository, then I think it would be best to do:

```
import fastcov
from unidiff import PatchSet
```

That way you don't need to do much testing because fastcov and unidiff are both independently tested.

If you decide to include this functionality in `fastcov.py` itself, then I think it is best to use the parsing class to avoid needing 3rd party dependencies in order to run core fastcov.

I'll leave it up to you to decide which you would rather use/maintain

rburtnykx commented:
As for me, I would like to put it info main fastcov.py. At least that is how we use it, just run single file without any dependencies. So i will prepare UT and functional tests for that logic

RPGillespie6 commented on test/unit/diff_tests_data/renamed_edited.diff:
```diff
@@ -0,0 +1,16 @@
++    assert not result added line
+
+ def test_DiffParser_parseDiffFile_deleted_file():
+     result = fastcov.DiffParser().parseDiffFile('diff_files/deleted_file.diff', '/base')
```
Very thorough number of unit test cases, I like it

rburtnykx replied:
Thanks, just tried to decrease risk of homegrown parser :)

RPGillespie6 commented on test/unit/test_diff_filter.py:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
+"""
+    Author: Bryan Gillespie
```
You can change the comment blocks of the unit tests you wrote to reflect that you are the author (if you wish). I don't want to take credit for someone else's work, though the commit history will show you as the true author.

rburtnykx replied:
Thanks. Basically, it's not important to me, but maybe someone will have some questions here or someone will find this code awful :) So I'll change it

RPGillespie6 commented on fastcov.py:
```diff
@@ -60,6 +60,133 @@ def format(self, record):
+            #sources without added lines will be excluded
+            if value:
+                newpath = os.path.join(diff_base_dir, key) if diff_base_dir else os.path.abspath(key)
+                newpath = newpath.replace('\\', '/')
```
Can you give an example where backslashes in a path would need to be converted to forward slashes?

rburtnykx replied:
That was some kind of cross-platform work with diff file filtering, but while I was writing unittest on Windows, completely forgot that this tool is not designed to be run on windows. So I'm deleting it

RPGillespie6 commented:
I'm going to go ahead and merge -- thanks for the contribution and supporting tests

rburtnykx commented:
Great, thanks for taking my changes
